pkg/cli/tail.go

```diff
@@ -222,24 +267,68 @@ func (c *TailCommand) Run(ctx context.Context, args []string) error {
 	return nil
 }
 
-func (c *TailCommand) tail(ctx context.Context) ([]*loggingpb.LogEntry, error) {
-	var p logPuller
-	if c.testPuller != nil {
-		p = c.testPuller
-	} else {
-		logClient, err := logging.NewClient(ctx)
-		if err != nil {
-			return nil, fmt.Errorf("failed to create logging client: %w", err)
+func (c *TailCommand) streamTail(ctx context.Context, extra []validation.Validator, puller logPuller) error {
+	// logCh is used to receive logEntriesfrom StreamPull
+	logCh := make(chan *loggingpb.LogEntry)
+	// errCh is used to receive error from StreamPull
+	errCh := make(chan error)
+	// doneCh is used to notifiy the main goroutinue when
+	// StreamPull is finished or canceled.
+	doneCh := make(chan struct{})
+
+	var failCount int
+	var totalCount int
+
+	go func() {
+		// close doneCh to notify main go rountine
+		// that receving from StreamPull has finished.
+		defer close(doneCh)
+
+		for l := range logCh {
+			totalCount++
+			js, err := protojson.Marshal(l)
+			if err != nil {
+				c.Errf("failed to marshal log to json (InsertId: %q): %w", l.InsertId, err)
+				continue
+			}
+			c.Outf(stripSpaces(string(js)))
+			if c.flagValidate {
+				if err := validation.Validate(string(js), extra...); err != nil {
+					failCount++
+					c.Errf("failed to validate log (InsertId: %q): %w\n", l.InsertId, err)
+				} else {
+					c.Outf("Successfully validated log (InsertId: %q)\n", l.InsertId)
+				}
+			}
 		}
-		p = cloudlogging.NewPuller(ctx, logClient, c.flagScope)
-	}
+	}()
 
-	ls, err := p.Pull(ctx, c.queryFilter(), c.flagMaxNum)
-	if err != nil {
-		return nil, fmt.Errorf("failed to pull logs: %w", err)
+	go func() {
+		if err := puller.StreamPull(ctx, c.queryFilter(), logCh); err != nil {
+			errCh <- fmt.Errorf("StreamPull failed: %w", err)
+		}
+		close(logCh)
+		close(errCh)
+	}()
+
+	select {
+	case <-ctx.Done():
+		// main channel will block here until doneCh receives
+		// this way we can prevent the possible race condition
+		// that main channel reads from totalCount and failCount
```

Comment:
It is not just totalCount and failCount, c.Outf and stdout.String() and etc. will also race if not blocked properly. Also we should move `if c.flagValidate {
			c.Outf("Validation failed for %d logs (out of %d)", failCount, totalCount)
		}` to the end of the read logCh goroutine. Doing so to avoid dup in line 327, they are out of the data race concern as they are only referenced there.

Reply:
Updated the comment.